File: redis/cpp/sender_async_test.cpp
#include <hiredis/hiredis.h>
#include <iostream>
#include <fstream>
#include <string>
#include <future>
#include <vector>
#include <thread>
#include <cstring>
#include "../../utils/cpp/stats_collector.hpp"
#include "../../utils/cpp/test_data_loader.hpp"
#include "../../utils/cpp/message_helpers.hpp"

using json = nlohmann::json;
using messaging::MessageEnvelope;
using message_helpers::get_current_time_ms;

struct TaskResult {
    bool success;
    std::string message_id;
    long long duration;
    std::string error;
};

TaskResult send_message_task(const json& item) {
    TaskResult res;
    res.success = false;
    res.message_id = message_helpers::extract_message_id(item);
    res.duration = 0;
    
    redisContext *c_pub = redisConnect("127.0.0.1", 6379);
    redisContext *c_sub = redisConnect("127.0.0.1", 6379);
    
    if (!c_pub || c_pub->err || !c_sub || c_sub->err) {
        res.error = "Connection failed";
        if (c_pub) redisFree(c_pub);
        if (c_sub) redisFree(c_sub);
        return res;
    }
    
    int target = item.value("target", 0);
    std::string channel = "test_channel_" + std::to_string(target);
    std::string reply_channel = "reply_" + res.message_id;
    
    // Subscribe to reply channel
    redisReply *sub = (redisReply*)redisCommand(c_sub, "SUBSCRIBE %s", reply_channel.c_str());
    if (sub) freeReplyObject(sub);
    
    // Send message
    long long msg_start = get_current_time_ms();
    MessageEnvelope envelope = message_helpers::create_data_envelope(item);
    (*envelope.mutable_metadata())["reply_to"] = reply_channel;
    std::string body = message_helpers::serialize_envelope(envelope);
    
    // Publish with retry to handle race condition where subscriber isn't ready
    int published_to = 0;
    for (int retry = 0; retry < 5 && published_to == 0; ++retry) {
        redisReply *pub = (redisReply*)redisCommand(c_pub, "PUBLISH %s %b", channel.c_str(), body.data(), body.size());
        if (pub) {
            if (pub->type == REDIS_REPLY_INTEGER) {
                published_to = (int)pub->integer;
            }
            freeReplyObject(pub);
        }
        if (published_to == 0) std::this_thread::sleep_for(std::chrono::milliseconds(2));
    }
    
    // Wait for ACK
    long long timeout_ms = 80;
    struct timeval tv;
    tv.tv_sec = 0;
    tv.tv_usec = timeout_ms * 1000;
    redisSetTimeout(c_sub, tv);
    
    while ((get_current_time_ms() - msg_start) < timeout_ms) {
        redisReply *reply = nullptr;
        int status = redisGetReply(c_sub, (void**)&reply);
        
        if (status == REDIS_OK && reply) {
            if (reply->type == REDIS_REPLY_ARRAY && reply->elements >= 3) {
                if (reply->element[0]->type == REDIS_REPLY_STRING && 
                    strcmp(reply->element[0]->str, "message") == 0) {
                    std::string reply_str(reply->element[2]->str, reply->element[2]->len);
                    
                    MessageEnvelope resp_envelope;
                    if (message_helpers::parse_envelope(reply_str, resp_envelope) && 
                        message_helpers::is_valid_ack(resp_envelope, res.message_id)) {
                        res.duration = get_current_time_ms() - msg_start;
                        res.success = true;
                        freeReplyObject(reply);
                        break;
                    }
                }
            }
            freeReplyObject(reply);
        } else if (status == REDIS_ERR) {
            if (c_sub->err == REDIS_ERR_IO && (errno == EAGAIN || errno == EINTR || errno == 0)) {
                c_sub->err = 0;
                memset(c_sub->errstr, 0, sizeof(c_sub->errstr));
            }
            break;
        }
    }
    
    // Unsubscribe
    redisSetTimeout(c_sub, {1, 0});
    redisReply *unsub = (redisReply*)redisCommand(c_sub, "UNSUBSCRIBE %s", reply_channel.c_str());
    if (unsub) freeReplyObject(unsub);
    
    redisFree(c_pub);
    redisFree(c_sub);
    
    if (!res.success && res.error.empty()) {
        res.error = "Timeout";
    }
    
    return res;
}

int main() {
    auto test_data = test_data_loader::loadTestData();

    MessageStats stats;
    stats.set_metadata({
        {"service", "Redis"},
        {"language", "C++"},
        {"async", true}
    });
    long long start_time = get_current_time_ms();

    std::cout << " [x] Starting ASYNC transfer of " << test_data.size() << " messages..." << std::endl;

    std::vector<std::future<TaskResult>> futures;
    for (auto& item : test_data) {
        futures.push_back(std::async(std::launch::async, send_message_task, item));
    }

    for (auto& fut : futures) {
        TaskResult res = fut.get();
        if (res.success) {
            stats.record_message(true, res.duration);
            std::cout << " [OK] Message " << res.message_id << " acknowledged" << std::endl;
        } else {
            stats.record_message(false);
            std::cout << " [FAILED] Message " << res.message_id << ": " << res.error << std::endl;
        }
    }

    long long end_time = get_current_time_ms();
    stats.set_duration(start_time, end_time);
    
    json report = stats.get_stats();

    std::cout << "\nTest Results (ASYNC):" << std::endl;
    std::cout << "total_sent: " << stats.sent_count << std::endl;
    std::cout << "total_received: " << stats.received_count << std::endl;
    std::cout << "duration_ms: " << stats.get_duration_ms() << std::endl;

    std::ofstream rf("logs/report.txt", std::ios::app);
    if (rf.good()) {
        rf << report.dump() << std::endl;
        rf.close();
    }

    return 0;
}

File: redis/python/sender_async_test.py
#!/usr/bin/env python3
"""Redis Python Sender - Async"""
import sys
import json
import asyncio
import redis.asyncio as redis
from pathlib import Path

# Add utils to path
repo_root = Path(__file__).parent.parent.parent
sys.path.insert(0, str(repo_root / 'utils' / 'python'))

from message_helpers import *
from test_data_loader import load_test_data
from stats_collector import MessageStats


async def send_message_task(item):
    """Send a single message asynchronously with its own connection."""
    result = {'success': False, 'message_id': '', 'duration': 0, 'error': ''}
    
    # Each task gets its own Redis connection
    r = None
    pubsub = None
    
    try:
        message_id = extract_message_id(item)
        result['message_id'] = message_id
        target = item.get('target', 0)
        
        channel_name = f"test_channel_{target}"
        reply_channel = f"reply_channel_{message_id}"
        
        msg_start = get_current_time_ms()
        
        # Create dedicated connection for this task
        r = redis.Redis(host='localhost', port=6379, db=0)
        pubsub = r.pubsub()
        
        # Subscribe to reply channel
        await pubsub.subscribe(reply_channel)
        
        # Create and send message
        envelope = create_data_envelope(item)
        envelope.metadata['reply_to'] = reply_channel
        body = serialize_envelope(envelope)
        
        # Publish
        await r.publish(channel_name, body)
        
        # Wait for reply with timeout
        try:
            while True:
                message = await pubsub.get_message(ignore_subscribe_messages=True, timeout=0.01)
                if message and message['type'] == 'message':
                    try:
                        resp_envelope = parse_envelope(message['data'])
                        if is_valid_ack(resp_envelope, message_id):
                            result['duration'] = get_current_time_ms() - msg_start
                            result['success'] = True
                            break
                        else:
                            # Ignore mismatched ACKs (usually late ACKs from previous messages)
                            pass
                    except Exception:
                        pass
                    
                # Timeout check (200ms - doubled for reliability)
                if (get_current_time_ms() - msg_start) > 200:
                    result['error'] = 'Timeout'
                    break
        except Exception as e:
            result['error'] = str(e)
            
    except Exception as e:
        result['error'] = str(e)
    finally:
        # Clean up resources
        if pubsub:
            try:
                await pubsub.unsubscribe(reply_channel)
                await pubsub.aclose()
            except:
                pass
        if r:
            try:
                await r.aclose()
            except:
                pass
    
    return result


async def run():
    test_data = load_test_data()
    
    stats = MessageStats()
    stats.set_metadata({
        'service': 'Redis',
        'language': 'Python',
        'async': True
    })
    start_time = get_current_time_ms()
    
    print(f" [x] Starting ASYNC transfer of {len(test_data)} messages...")
    
    # Give receivers time to subscribe (Redis pub/sub doesn't queue messages)
    await asyncio.sleep(0.5)
    
    # Process concurrently - each task gets its own connection
    tasks = [send_message_task(item) for item in test_data]
    results = await asyncio.gather(*tasks)
    
    for result in results:
        if result['success']:
            stats.record_message(True, result['duration'])
            print(f" [OK] Message {result['message_id']} acknowledged")
        else:
            stats.record_message(False)
            print(f" [FAILED] Message {result['message_id']}: {result['error']}")
    
    end_time = get_current_time_ms()
    stats.set_duration(start_time, end_time)
    
    report = stats.get_stats()
    
    print("\nTest Results (ASYNC):")
    print(f"total_sent: {stats.sent_count}")
    print(f"total_received: {stats.received_count}")
    print(f"duration_ms: {stats.get_duration_ms()}")
    
    with open('logs/report.txt', 'a') as f:
        f.write(json.dumps(report) + '\n')


def main():
    asyncio.run(run())


if __name__ == "__main__":
    main()
